### ISSUE TITLE
Bug: `no-useless-assignment` false positive

### ISSUE THREAD

**Issue by gian1200** [eslint/eslint#20609]
### Environment

Node version: v24.12.0
npm version: 11.10.1
Local ESLint version: 10.0.3
Global ESLint version: N/A
Operating System: Windows 11

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
import globals from "globals";

import js from "@eslint/js";
import stylistic from "@stylistic/eslint-plugin";
import jsdoc from "eslint-plugin-jsdoc";

/**
 * @type {import("eslint").Linter.Config}
 */
export default [
	{
		"ignores": [".git", ".husky", ".scannerwork", ".vscode", "coverage", "dependency-check-bin", "dependency-check-reports", "dist", "node_modules", "target"],
		"name": "Archivos a ignorar"
	},
	{
		"files": ["**/*.{js,cjs,mjs}"],
		"languageOptions": {
			"ecmaVersion": 2024
		},
		"name": "Archivos JavaScript",
		"plugins": {
			"@stylistic": stylistic,
			"jsdoc": jsdoc
		},
		"rules": {
			...js.configs.recommended.rules,
			...jsdoc.configs["flat/recommended-typescript-flavor"].rules,
			"@stylistic/array-bracket-newline": ["error", "consistent"],
			"@stylistic/array-bracket-spacing": ["error", "never"],
			"@stylistic/array-element-newline": ["error", { "ArrayExpression": "consistent" }],
			"@stylistic/arrow-spacing": ["warn", { "before": true, "after": true }],
			"@stylistic/block-spacing": ["warn", "always"],
			"@stylistic/brace-style": ["warn", "1tbs", { "allowSingleLine": false }],
			"@stylistic/comma-dangle": ["error", "never"],
			"@stylistic/comma-spacing": ["warn", { "before": false, "after": true }],
			"@stylistic/eol-last": ["error"],
			"@stylistic/function-call-spacing": ["error", "never"],
			"@stylistic/indent": ["warn", "tab", { "SwitchCase": 1 }],
			"@stylistic/key-spacing": ["warn", { "afterColon": true }],
			"@stylistic/keyword-spacing": ["warn"],
			"@stylistic/linebreak-style": ["off"],
			"@stylistic/lines-around-comment": ["error", { "allowBlockStart": true, "allowClassStart": false }],
			"@stylistic/lines-between-class-members": ["warn", "always"],
			"@stylistic/no-multi-spaces": ["warn"],
			"@stylistic/no-multiple-empty-lines": ["warn", { "max": 1, "maxEOF": 0 }],
			"@stylistic/no-trailing-spaces": ["warn"],
			"@stylistic/nonblock-statement-body-position": ["error", "below"],
			"@stylistic/object-curly-newline": ["error", { "consistent": true }],
			"@stylistic/object-curly-spacing": ["warn", "always"],
			"@stylistic/object-property-newline": ["error", { "allowAllPropertiesOnSameLine": true }],
			"@stylistic/padded-blocks": ["warn", { "blocks": "never", "classes": "always", "switches": "never" }, { "allowSingleLineBlocks": false }],
			"@stylistic/quote-props": ["warn"],
			"@stylistic/quotes": ["warn", "double", { "allowTemplateLiterals": "always", "avoidEscape": true }],
			"@stylistic/semi": ["warn", "always"],
			"@stylistic/space-before-blocks": ["warn"],
			"@stylistic/space-infix-ops": ["warn"],
			"@stylistic/space-in-parens": ["warn", "never"],
			"array-callback-return": ["error"],
			"curly": ["warn", "all"],
			"default-param-last": ["error"],
			"eqeqeq": ["warn", "smart"],
			"grouped-accessor-pairs": ["error", "getBeforeSet"],
			"no-console": ["warn"],
			"no-debugger": ["warn"],
			"no-empty": ["warn"],
			"no-lonely-if": ["warn"],
			"no-multi-str": ["error"],
			"no-return-assign": ["error"],
			"no-shadow": ["warn", { "builtinGlobals": true, "hoist": "all" }],
			"no-unneeded-ternary": ["error"],
			"no-unused-vars": ["warn"],
			"no-var": ["error"],
			"prefer-const": ["warn"]
		}
	},
	{
		"files": ["**/*.mjs"],
		"name": "Módulos JavaScript",
		"languageOptions": {
			"sourceType": "module"
		}
	},
	{
		"files": ["src/app/**/*.{js,mjs}"],
		"name": "Archivos web",
		"languageOptions": {
			"ecmaVersion": 2022,
			"globals": {
				...globals.browser,
				...globals.jquery,
				"Chart": "readonly",
				"DataTable": "readable",
				"moment": "readonly",
				"saveAs": "readonly"
			}
		}
	},
	{
		"files": [".config/karma.conf.cjs", "**/*.{cjs,mjs}"],
		"name": "Archivos Node.js",
		"ignores": ["src/app/**/*"],
		"languageOptions": {
			"globals": {
				...globals.node
			}
		}
	},
	{
		"files": ["src/app/**/*.spec.*"],
		"name": "Pruebas web",
		"languageOptions": {
			"globals": {
				...globals.jasmine
			}
		}
	}
];

```
</details>

```js
const anArray = [];

const test = (aaa) => {
	let strCategoryForm = "";

	if (aaa?.length === 0) {
		strCategoryForm = "any String A";
	} else {
		strCategoryForm = "any String B";
	}

	anArray.push({
		"action": {
			"category": strCategoryForm
		},
		"event": "trackAction"
	});
};
```

### What did you expect to happen?

No error

### What actually happened?

The value assigned to 'strCategoryForm' is not used in subsequent statements.

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiY29uc3QgYW5BcnJheSA9IFtdO1xuXG5jb25zdCB0ZXN0ID0gKGFhYSkgPT4ge1xuXHRsZXQgc3RyQ2F0ZWdvcnlGb3JtID0gXCJcIjtcblxuXHRpZiAoYWFhPy5sZW5ndGggPT09IDApIHtcblx0XHRzdHJDYXRlZ29yeUZvcm0gPSBcImFueSBTdHJpbmcgQVwiO1xuXHR9IGVsc2Uge1xuXHRcdHN0ckNhdGVnb3J5Rm9ybSA9IFwiYW55IFN0cmluZyBCXCI7XG5cdH1cblxuXHRhbkFycmF5LnB1c2goe1xuXHRcdFwiYWN0aW9uXCI6IHtcblx0XHRcdFwiY2F0ZWdvcnlcIjogc3RyQ2F0ZWdvcnlGb3JtXG5cdFx0fSxcblx0XHRcImV2ZW50XCI6IFwidHJhY2tBY3Rpb25cIlxuXHR9KTtcbn07XG4iLCJvcHRpb25zIjp7InJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjpbImVycm9yIl0sImZvci1kaXJlY3Rpb24iOlsiZXJyb3IiXSwiZ2V0dGVyLXJldHVybiI6WyJlcnJvciJdLCJuby1hc3luYy1wcm9taXNlLWV4ZWN1dG9yIjpbImVycm9yIl0sIm5vLWNhc2UtZGVjbGFyYXRpb25zIjpbImVycm9yIl0sIm5vLWNsYXNzLWFzc2lnbiI6WyJlcnJvciJdLCJuby1jb21wYXJlLW5lZy16ZXJvIjpbImVycm9yIl0sIm5vLWNvbmQtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbnN0LWFzc2lnbiI6WyJlcnJvciJdLCJuby1jb25zdGFudC1iaW5hcnktZXhwcmVzc2lvbiI6WyJlcnJvciJdLCJuby1jb25zdGFudC1jb25kaXRpb24iOlsiZXJyb3IiXSwibm8tY29udHJvbC1yZWdleCI6WyJlcnJvciJdLCJuby1kZWJ1Z2dlciI6WyJlcnJvciJdLCJuby1kZWxldGUtdmFyIjpbImVycm9yIl0sIm5vLWR1cGUtYXJncyI6WyJlcnJvciJdLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOlsiZXJyb3IiXSwibm8tZHVwZS1lbHNlLWlmIjpbImVycm9yIl0sIm5vLWR1cGUta2V5cyI6WyJlcnJvciJdLCJuby1kdXBsaWNhdGUtY2FzZSI6WyJlcnJvciJdLCJuby1lbXB0eSI6WyJlcnJvciJdLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOlsiZXJyb3IiXSwibm8tZW1wdHktcGF0dGVybiI6WyJlcnJvciJdLCJuby1lbXB0eS1zdGF0aWMtYmxvY2siOlsiZXJyb3IiXSwibm8tZXgtYXNzaWduIjpbImVycm9yIl0sIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6WyJlcnJvciJdLCJuby1mYWxsdGhyb3VnaCI6WyJlcnJvciJdLCJuby1mdW5jLWFzc2lnbiI6WyJlcnJvciJdLCJuby1nbG9iYWwtYXNzaWduIjpbImVycm9yIl0sIm5vLWltcG9ydC1hc3NpZ24iOlsiZXJyb3IiXSwibm8taW52YWxpZC1yZWdleHAiOlsiZXJyb3IiXSwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOlsiZXJyb3IiXSwibm8tbG9zcy1vZi1wcmVjaXNpb24iOlsiZXJyb3IiXSwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOlsiZXJyb3IiXSwibm8tbmV3LW5hdGl2ZS1ub25jb25zdHJ1Y3RvciI6WyJlcnJvciJdLCJuby1ub25vY3RhbC1kZWNpbWFsLWVzY2FwZSI6WyJlcnJvciJdLCJuby1vYmotY2FsbHMiOlsiZXJyb3IiXSwibm8tb2N0YWwiOlsiZXJyb3IiXSwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjpbImVycm9yIl0sIm5vLXJlZGVjbGFyZSI6WyJlcnJvciJdLCJuby1yZWdleC1zcGFjZXMiOlsiZXJyb3IiXSwibm8tc2VsZi1hc3NpZ24iOlsiZXJyb3IiXSwibm8tc2V0dGVyLXJldHVybiI6WyJlcnJvciJdLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6WyJlcnJvciJdLCJuby1zcGFyc2UtYXJyYXlzIjpbImVycm9yIl0sIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjpbImVycm9yIl0sIm5vLXVuYXNzaWduZWQtdmFycyI6WyJlcnJvciJdLCJuby11bmRlZiI6WyJlcnJvciJdLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6WyJlcnJvciJdLCJuby11bnJlYWNoYWJsZSI6WyJlcnJvciJdLCJuby11bnNhZmUtZmluYWxseSI6WyJlcnJvciJdLCJuby11bnNhZmUtbmVnYXRpb24iOlsiZXJyb3IiXSwibm8tdW5zYWZlLW9wdGlvbmFsLWNoYWluaW5nIjpbImVycm9yIl0sIm5vLXVudXNlZC1sYWJlbHMiOlsiZXJyb3IiXSwibm8tdW51c2VkLXByaXZhdGUtY2xhc3MtbWVtYmVycyI6WyJlcnJvciJdLCJuby11bnVzZWQtdmFycyI6WyJlcnJvciJdLCJuby11c2VsZXNzLWFzc2lnbm1lbnQiOlsiZXJyb3IiXSwibm8tdXNlbGVzcy1iYWNrcmVmZXJlbmNlIjpbImVycm9yIl0sIm5vLXVzZWxlc3MtY2F0Y2giOlsiZXJyb3IiXSwibm8tdXNlbGVzcy1lc2NhcGUiOlsiZXJyb3IiXSwibm8td2l0aCI6WyJlcnJvciJdLCJwcmVzZXJ2ZS1jYXVnaHQtZXJyb3IiOlsiZXJyb3IiXSwicmVxdWlyZS15aWVsZCI6WyJlcnJvciJdLCJ1c2UtaXNuYW4iOlsiZXJyb3IiXSwidmFsaWQtdHlwZW9mIjpbImVycm9yIl19LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==

### Participation

- [ ] I am willing to submit a pull request for this issue.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

**Comment by DMartens** [comment]
The rule is correct in this case as the variable `strCategoriesForm` is not read before the `if/else` where the value is overwritten.
As such the initial value (empty string) is useless.

**Comment by gian1200** [comment]
But it is read after the if/else. That part is still within the same scope.

- If there is only "if", it works.
- It even work if there is "if/else-if".
 
It seems to fail only when the variable is updated on each scenario. If at least 1 scenario do not edit the value, then the rule is not triggered.

[Updated reproducible example](https://eslint.org/play/#eyJ0ZXh0IjoiY29uc3QgYW5BcnJheSA9IFtdO1xuXG5jb25zdCB0ZXN0ID0gKGFhYSkgPT4ge1xuXHRsZXQgc3RyQ2F0ZWdvcnlGb3JtID0gXCJcIjtcblxuXHRpZiAoYWFhPy5sZW5ndGggPT09IDApIHtcblx0XHRzdHJDYXRlZ29yeUZvcm0gPSBcImFueSBTdHJpbmcgQVwiO1xuXHR9IGVsc2UgaWYgKGFhYT8ubGVuZ3RoID09PSAxKSB7XG5cdFx0Ly8gc3RyQ2F0ZWdvcnlGb3JtID0gXCJhbnkgU3RyaW5nIEJcIjtcblx0fSBlbHNlIHtcblx0XHRzdHJDYXRlZ29yeUZvcm0gPSBcImFueSBTdHJpbmcgQ1wiO1xuICAgIH1cbiAgXG5cdGFuQXJyYXkucHVzaCh7XG5cdFx0XCJhY3Rpb25cIjoge1xuXHRcdFx0XCJjYXRlZ29yeVwiOiBzdHJDYXRlZ29yeUZvcm1cblx0XHR9LFxuXHRcdFwiZXZlbnRcIjogXCJ0cmFja0FjdGlvblwiXG5cdH0pO1xufTtcbiIsIm9wdGlvbnMiOnsicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOlsiZXJyb3IiXSwiZm9yLWRpcmVjdGlvbiI6WyJlcnJvciJdLCJnZXR0ZXItcmV0dXJuIjpbImVycm9yIl0sIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOlsiZXJyb3IiXSwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOlsiZXJyb3IiXSwibm8tY2xhc3MtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbXBhcmUtbmVnLXplcm8iOlsiZXJyb3IiXSwibm8tY29uZC1hc3NpZ24iOlsiZXJyb3IiXSwibm8tY29uc3QtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbnN0YW50LWJpbmFyeS1leHByZXNzaW9uIjpbImVycm9yIl0sIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6WyJlcnJvciJdLCJuby1jb250cm9sLXJlZ2V4IjpbImVycm9yIl0sIm5vLWRlYnVnZ2VyIjpbImVycm9yIl0sIm5vLWRlbGV0ZS12YXIiOlsiZXJyb3IiXSwibm8tZHVwZS1hcmdzIjpbImVycm9yIl0sIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6WyJlcnJvciJdLCJuby1kdXBlLWVsc2UtaWYiOlsiZXJyb3IiXSwibm8tZHVwZS1rZXlzIjpbImVycm9yIl0sIm5vLWR1cGxpY2F0ZS1jYXNlIjpbImVycm9yIl0sIm5vLWVtcHR5IjpbImVycm9yIl0sIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6WyJlcnJvciJdLCJuby1lbXB0eS1wYXR0ZXJuIjpbImVycm9yIl0sIm5vLWVtcHR5LXN0YXRpYy1ibG9jayI6WyJlcnJvciJdLCJuby1leC1hc3NpZ24iOlsiZXJyb3IiXSwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjpbImVycm9yIl0sIm5vLWZhbGx0aHJvdWdoIjpbImVycm9yIl0sIm5vLWZ1bmMtYXNzaWduIjpbImVycm9yIl0sIm5vLWdsb2JhbC1hc3NpZ24iOlsiZXJyb3IiXSwibm8taW1wb3J0LWFzc2lnbiI6WyJlcnJvciJdLCJuby1pbnZhbGlkLXJlZ2V4cCI6WyJlcnJvciJdLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6WyJlcnJvciJdLCJuby1sb3NzLW9mLXByZWNpc2lvbiI6WyJlcnJvciJdLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6WyJlcnJvciJdLCJuby1uZXctbmF0aXZlLW5vbmNvbnN0cnVjdG9yIjpbImVycm9yIl0sIm5vLW5vbm9jdGFsLWRlY2ltYWwtZXNjYXBlIjpbImVycm9yIl0sIm5vLW9iai1jYWxscyI6WyJlcnJvciJdLCJuby1vY3RhbCI6WyJlcnJvciJdLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOlsiZXJyb3IiXSwibm8tcmVkZWNsYXJlIjpbImVycm9yIl0sIm5vLXJlZ2V4LXNwYWNlcyI6WyJlcnJvciJdLCJuby1zZWxmLWFzc2lnbiI6WyJlcnJvciJdLCJuby1zZXR0ZXItcmV0dXJuIjpbImVycm9yIl0sIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjpbImVycm9yIl0sIm5vLXNwYXJzZS1hcnJheXMiOlsiZXJyb3IiXSwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOlsiZXJyb3IiXSwibm8tdW5hc3NpZ25lZC12YXJzIjpbImVycm9yIl0sIm5vLXVuZGVmIjpbImVycm9yIl0sIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjpbImVycm9yIl0sIm5vLXVucmVhY2hhYmxlIjpbImVycm9yIl0sIm5vLXVuc2FmZS1maW5hbGx5IjpbImVycm9yIl0sIm5vLXVuc2FmZS1uZWdhdGlvbiI6WyJlcnJvciJdLCJuby11bnNhZmUtb3B0aW9uYWwtY2hhaW5pbmciOlsiZXJyb3IiXSwibm8tdW51c2VkLWxhYmVscyI6WyJlcnJvciJdLCJuby11bnVzZWQtcHJpdmF0ZS1jbGFzcy1tZW1iZXJzIjpbImVycm9yIl0sIm5vLXVudXNlZC12YXJzIjpbImVycm9yIl0sIm5vLXVzZWxlc3MtYXNzaWdubWVudCI6WyJlcnJvciJdLCJuby11c2VsZXNzLWJhY2tyZWZlcmVuY2UiOlsiZXJyb3IiXSwibm8tdXNlbGVzcy1jYXRjaCI6WyJlcnJvciJdLCJuby11c2VsZXNzLWVzY2FwZSI6WyJlcnJvciJdLCJuby13aXRoIjpbImVycm9yIl0sInByZXNlcnZlLWNhdWdodC1lcnJvciI6WyJlcnJvciJdLCJyZXF1aXJlLXlpZWxkIjpbImVycm9yIl0sInVzZS1pc25hbiI6WyJlcnJvciJdLCJ2YWxpZC10eXBlb2YiOlsiZXJyb3IiXX0sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19)


When line 9 or line 11 are commented, the rule is not triggered. Is this by design?
